python/cudf/cudf/core/column/column.py

```diff
@@ -2033,69 +2033,121 @@ def as_column(
+            )
+            or (
+                isinstance(arbitrary.dtype, pd.CategoricalDtype)
+                and isinstance(
+                    arbitrary.dtype.categories.dtype, pd.DatetimeTZDtype
                 )
-            data = as_column(pa.array(arbitrary, from_pandas=True))
-        elif arbitrary.dtype == np.bool_:
-            data = as_column(cupy.asarray(arbitrary), dtype=arbitrary.dtype)
-        elif arbitrary.dtype.kind in ("f"):
-            arb_dtype = np.dtype(arbitrary.dtype)
+            )
+        ):
+            raise NotImplementedError(
+                "cuDF does not yet support timezone-aware datetimes"
+            )
+        elif _is_pandas_nullable_extension_dtype(arbitrary.dtype):
+            if cudf.get_option("mode.pandas_compatible"):
+                raise NotImplementedError("not supported")
+            if isinstance(arbitrary, (pd.Series, pd.Index)):
+                arbitrary = arbitrary.array
             data = as_column(
-                cupy.asarray(arbitrary, dtype=arb_dtype),
+                pa.array(arbitrary, from_pandas=True),
                 nan_as_null=nan_as_null,
                 dtype=dtype,
+                length=length,
             )
-        elif arbitrary.dtype.kind in ("u", "i"):
+        elif isinstance(
+            arbitrary.dtype, (pd.CategoricalDtype, pd.IntervalDtype)
+        ):
+            # TODO: Validate subtype?
             data = as_column(
-                cupy.asarray(arbitrary), nan_as_null=nan_as_null, dtype=dtype
+                pa.array(arbitrary, from_pandas=True),
+                nan_as_null=nan_as_null,
+                dtype=dtype,
+                length=length,
             )
-        elif isinstance(arbitrary.dtype, pd.PeriodDtype):
+        elif isinstance(
+            arbitrary.dtype, pd.api.extensions.ExtensionDtype
+        ) and not isinstance(arbitrary, pd.arrays.PandasArray):
             raise NotImplementedError(
-                "cuDF does not yet support `PeriodDtype`"
+                "Custom pandas ExtensionDtypes are not supported"
             )
-        else:
-            if cudf.get_option(
-                "mode.pandas_compatible"
-            ) and _is_pandas_nullable_extension_dtype(arbitrary.dtype):
-                raise NotImplementedError("not supported")
-            pyarrow_array = pa.array(arbitrary, from_pandas=nan_as_null)
-            if arbitrary.dtype == cudf.dtype("object") and cudf.dtype(
-                pyarrow_array.type.to_pandas_dtype()
-            ) != cudf.dtype(arbitrary.dtype):
+        elif arbitrary.dtype.kind in "fiubmM":
+            if isinstance(arbitrary, pd.arrays.PandasArray):
+                arbitrary = np.array(arbitrary)
+            arb_dtype = np.dtype(arbitrary.dtype)
+            if arb_dtype.kind == "f" and arb_dtype.itemsize == 2:
+                # float16
+                arbitrary = arbitrary.astype(np.dtype(np.float32))
```

Comment:
Yeah I believe so I was migrating this code below here

```python
               arb_dtype = (
                    cudf.dtype("float32")
                    if arbitrary.dtype == "float16"
                    else cudf.dtype(arbitrary.dtype)
                )
```

pandas also does not support `float16` (which was made more intentional by raising in pandas 2.0, before it would sometimes coerce to `float32` also), so I don't think the dtype preservation here isn't too much of an issue

Reply:
Ah so my generalization about float16 in pandas 2.0 isn't entirely correct.

Only pandas Index objects will disallow float16 (IIRC there's no hashtable implementation for this type), but Series and DataFrame objects will continue to allow float16.

Reply:
Yeah agreed raising here is better than silent upcasting, I'll make this raise and mark as a breaking change